### ISSUE TITLE
core sometimes cannot publish to history

### ISSUE THREAD

**Issue by MonsieurNicolas** [stellar/stellar-core#1746]
# Issue type
  bug

# Version found
9.2.0

# Issue description
We found some validators had a publish backlog, core was in sync but not trying to publish (it appeared).

## Actual result
For some reason publish was not happening. The number of checkpoints to publish was in the order of 400 on one of the machines (over 30 hours).

A possible theory is that something was blocking the main thread long enough to timeout publish before it had a chance to perform work.

We need to investigate both:
- [ ] why publish would stop working (starting with its timeout logic)
- [ ] the reason why the bucket GC code was so slow

## Expected result
 core should always publish its backlog.

## supporting logs

looks like deleting files took a long time as per logs.

CPU usage was very high when there was a backlog; on one of the machines, the CPU usage went back down as soon as publish completed consuming its backlog.

```
2018-07-26T10:47:52.801 GCGB2 [Ledger INFO] Got consensus: [seq=19125605, prev=bcbb7d, tx_count=9, sv: [  txH: 130a9f, ct: 1532602070, upgrades: [ ] ]]
2018-07-26T10:47:52.801 GCGB2 [Ledger DEBUG] starting closeLedger() on ledgerSeq=19125605
2018-07-26T10:47:52.802 GCGB2 [Ledger DEBUG] processing fees and sequence numbers
2018-07-26T10:47:52.815 GCGB2 [Tx DEBUG] applyTransactions: ledger = 19125605
2018-07-26T10:47:52.815 GCGB2 [Tx DEBUG]  tx#0 = 0b96bd txseq=81733562650380303 (@ GBNZJ)
2018-07-26T10:47:52.818 GCGB2 [Tx DEBUG]  tx#1 = 77f698 txseq=80366504624787546 (@ GBHXT)
2018-07-26T10:47:52.821 GCGB2 [Tx DEBUG]  tx#2 = 8a35d5 txseq=75187504210775890 (@ GDD7A)
2018-07-26T10:47:52.846 GCGB2 [Tx DEBUG]  tx#3 = b800a1 txseq=78090588569860497 (@ GBGPP)
2018-07-26T10:47:52.848 GCGB2 [Tx DEBUG]  tx#4 = a862e0 txseq=78090472605707320 (@ GDS2O)
2018-07-26T10:47:52.949 GCGB2 [Tx DEBUG]  tx#5 = aebaa5 txseq=77833036561751526 (@ GAOVL)
2018-07-26T10:47:52.960 GCGB2 [Tx DEBUG]  tx#6 = dbd1ca txseq=80006792523808780 (@ GBG2T)
2018-07-26T10:47:52.966 GCGB2 [Tx DEBUG]  tx#7 = f630e6 txseq=75110564166107403 (@ GBNMQ)
2018-07-26T10:47:52.970 GCGB2 [Bucket DEBUG] Deleting bucket file /data/buckets/tmp/bucket-22d271cafed383f1/tmp-bucket-132a233d92f913b5.xdr that is redundant with existing bucke
t
2018-07-26T10:47:52.973 GCGB2 [Tx DEBUG]  tx#8 = e652ab txseq=82134978883747842 (@ GAKXV)
2018-07-26T10:47:52.980 GCGB2 [Bucket DEBUG] Adopting bucket file /data/buckets/tmp/bucket-22d271cafed383f1/tmp-bucket-21f0f5d7ffb7fdad.xdr as /data/buckets/bucket-7f218b08c54cf181747d7a22dfb2f152546041c1e8e9a8d5d3daae30bfe2456f.xdr
2018-07-26T10:47:52.980 GCGB2 [Bucket DEBUG] Adopting bucket file /data/buckets/tmp/bucket-22d271cafed383f1/tmp-bucket-9bee6c206afc3db3.xdr as /data/buckets/bucket-d682bb48ecaf3c39ee8bd534523b44ca25b9eef9a76363d4842e5ba90f9659b4.xdr
2018-07-26T10:47:52.981 GCGB2 [Bucket DEBUG] Adopting bucket file /data/buckets/tmp/bucket-22d271cafed383f1/tmp-bucket-b1c35543e4957381.xdr as /data/buckets/bucket-902f832d0cbe48d79af52c9364596a9fa37f82596ec02949bfd0fdbd5e65e724.xdr
2018-07-26T10:47:54.321 GCGB2 [Bucket DEBUG] Deleting bucket file /data/buckets/tmp/bucket-22d271cafed383f1/tmp-bucket-e838c9aafdf30655.xdr that is redundant with existing bucket
2018-07-26T10:47:54.436 GCGB2 [Bucket DEBUG] Deleting bucket file /data/buckets/tmp/bucket-22d271cafed383f1/tmp-bucket-293687b43e0de5d1.xdr that is redundant with existing bucket
2018-07-26T10:48:09.343 GCGB2 [Bucket DEBUG] Deleting bucket file /data/buckets/tmp/bucket-22d271cafed383f1/tmp-bucket-e894257d16983749.xdr that is redundant with existing bucket
2018-07-26T10:48:09.363 GCGB2 [Bucket DEBUG] Adopting bucket file /data/buckets/tmp/bucket-22d271cafed383f1/tmp-bucket-1f96bd5cb652d7a7.xdr as /data/buckets/bucket-2be6f3e05cde2181f4555a992da14f7de751688ad883dedb6653bba7d0c9808d.xdr
2018-07-26T10:48:09.366 GCGB2 [Ledger DEBUG] Advancing LCL: [seq=19125604, hash=bcbb7d] -> [seq=19125605, hash=ed1385]
2018-07-26T10:48:09.366 GCGB2 [Ledger DEBUG] New current ledger: seq=19125606
```

**Comment by MonsieurNicolas** [comment]
I opened #1764 to track the hang

**Comment by fuzzyami** [comment]
@MonsieurNicolas I think we just hit this issue on one of our networks. I'm seeing (1) backlog in publishing history checkpoints (2) deleting files in /bucket/tmp takes about 25 seconds.

This issue coincided with large fluctuations in free disc space on the Core machines. Seems like these files (in /buckets/tmp) are huge - 5Gb or so. Every ~10 minutes the core generates and then evicts these large files. This is definitely new. Perhaps we crossed some threshold in the bucket-list and it now generates huge files?

**Comment by MonsieurNicolas** [comment]
@fuzzyami what you are seeing is #1764.
This issue is tracking something else: core was not publishing even though it had many things to publish

**Comment by fuzzyami** [comment]
thank you for your response. I'm not sure why reached that conclusion - we
have exactly this symptom: erratic publishing behaviour, with the core
publishing every once in a while and mostly not publishing, thus
accumulating a long checkpoint queue. From what I gather this is just a
symptom of another issue - the main thread is too busy deleting large files
in buckets/tmp and so it misses its publishing window (because the lcl >
local ledger?). At any rate I'm curious about why this is happening to
begin with. Why are these files so big all of the sudden? and why does it
take so long to delete them? would switching to async deletion solve this?

**Comment by MonsieurNicolas** [comment]
maybe I misread your response, are you saying that publishing is extremely slow (that would maybe be related to #1764 - it causes the node to not close ledgers quickly as well) or not happening at all (this issue)?

**Comment by MonsieurNicolas** [comment]
@fuzzyami not sure if you see my response if I don't explicitly @ mention you

**Comment by fuzzyami** [comment]
I did :]
the publishing behaviour is erratic: sometimes every 10 minutes, sometimes every 50 - there's no method to this particular madness. As I said I *think* it has to do with huge temp bucket files suddenly  appearing on every core (only to be deleted a few minutes later), which I suspect is the root cause here. Why would I start seeing such huge files? could this be just the normal behaviour of the bucket list? Does it start generating large files once the total blockchain state crosses some threshold?

**Comment by MonsieurNicolas** [comment]
you should open a new issue for large buckets appearing: right now the largest bucket we have on the network is ~300MB at level 10 so it gets modified only once every two months.

As for the publish failing somehow, you mentioned that you think "because the lcl > local ledger?" - do you have any logs of interest that would help pinpoint the issue of publish not doing its job?

Most of the logging should show up with a verbosity of DEBUG under log lines labeled "Work"

**Comment by fuzzyami** [comment]
its pure guesswork at this stage. we'll share once we have something
tangible. thx again!

**Comment by fuzzyami** [comment]
@MonsieurNicolas I do have some updates from my side. 

As the number of queued checkpoints grew beyond 1K on our cores, we had decided to change the checkpoint-publishing code a little: we capped the back-off code to retry every ~10 seconds. Originally we expected this change to impact just the rate of publishing, but as it turned out, more happened.

As expected, our cores started publishing their queued checkpoints much faster. To our surprise, as the queues shrunk, it also favourably changed the behaviour of the core. Specifically, the stopped churning these large (5Gb) files every ~10 minutes or so. 

its a little hard to determine which is the cause and which is the effect in this case, but we originally thought that crossing into the 10th level of the bucket list caused our cores to merge slowly. This in turn could explain why our checkpoints were not publishing (because the core doesn't publish while merging).

We are no longer sure what caused what. Clearly, by reducing the checkpoint-queue we were able to return to "normal" core behaviour, so perhaps the story here is different. We're not sure what caused our merges to take forever. **At any rate I suggest you examine the behaviour of the core when encumbered by a long checkpoint queue, although I cant tell you what caused it to begin with**.

We deliberately left core in that state and we intend to continue to investigate this, should you have any ideas on what to look at.

**Comment by fuzzyami** [comment]
Oh, yeah, there's one more mystery here: 
As our cores started publishing their checkpoint more aggressively, we noticed 2 thing:
1. at some point, once the queue went below some threshold, the rate of publishing increased dramatically. this is another indication that having a large queue influences the rate of publishing.
2. however, we expected this threshold to be identical across the cores, and it wasn't. despite running the exact same code, the different cores exhibited different thresholds. see the graph I've attached.
